sklearn/model_selection/_search.py

```diff
@@ -520,6 +547,480 @@ def inverse_transform(self, Xt):
+                                                fit_params=self.fit_params,
+                                                error_score=self.error_score)
+            for parameters in parameter_iterable)
+
+        for score, parameters, _time in out:
+            grid_scores.append(_ScoreTuple(parameters, score))
+
+        # Store the computed scores
+        self.grid_scores_ = grid_scores
+
+        # Find the best parameters by comparing on the score:
+        # note that `sorted` is deterministic in the way it breaks ties
+        best = sorted(grid_scores, key=lambda x: x.score,
+                      reverse=True)[0]
+
+        self.best_params_ = best.parameters
+        self.best_score_ = best.score
+
+        if self.refit:
+            # fit the best estimator using the entire dataset
+            # clone first to work around broken estimators
+            best_estimator = clone(estimator).set_params(
+                **best.parameters)
+            if y is not None:
+                best_estimator.fit(X, y=None, **self.fit_params)
+            else:
+                best_estimator.fit(X, **self.fit_params)
+            self.best_estimator_ = best_estimator
+        return self
+
+
+def fit_and_score_unsupervised(estimator, X, scorer, verbose, parameters,
+                               fit_params, y=None, error_score='raise'):
+    """Fit estimator and compute scores for a given dataset.
+
+    Parameters
+    ----------
+    estimator : estimator object implementing 'fit'
+        The object to use to fit the data.
+
+    X : array-like of shape at least 2D
+        The data to fit.
+
+    y : array-like, shape = [n_samples] or [n_samples, n_output], optional
+            None for unsupervised learning.
+
+    scorer : callable
+        A scorer callable object / function with signature
+        ``scorer(estimator, X, y)`` with y being None for unsupervised learning.
+
+    verbose : integer
+        The verbosity level.
+
+    error_score : 'raise' (default) or numeric
+        Value to assign to the score if an error occurs in estimator fitting.
+        If set to 'raise', the error is raised. If a numeric value is given,
+        FitFailedWarning is raised. This parameter does not affect the refit
+        step, which will always raise the error.
+
+    parameters : dict or None
+        Parameters to be set on the estimator.
+
+    fit_params : dict or None
+        Parameters that will be passed to ``estimator.fit``.
+
+    Returns
+    -------
+    score : float
+        Score on the data to fit.
+
+    parameters : dict or None, optional
+        The parameters that have been evaluated.
+
+    scoring_time : float
+        Time spent for fitting and scoring in seconds.
+
+    """
+    if parameters is not None:
+        estimator.set_params(**parameters)
+
+    start_time = time.time()
+
+    try:
+        estimator.fit(X, y, **fit_params)
+        score = scorer(estimator, X)
```

Comment:
If unsupervised, `y` should not be provided to `fit`. But it may need to be provided to some scoring metrics. If semi-supervised, we're entering new territory.